azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/AzureAccount.java

```diff
@@ -45,128 +47,144 @@ public Account account() throws AzureToolkitAuthenticationException {
     }
 
-    public AzureAccount login(AuthType type) throws LoginFailureException {
-        AuthConfiguration auth = new AuthConfiguration();
-        auth.setType(type);
-        loginWithAuthConfiguration(auth);
-        return this;
+    public Mono<Account> authenticateAccounts(List<Account> accounts) {
+        Mono<Account> current = accounts.get(0).authenticate();
+        for (int i = 1; i < accounts.size(); i++) {
+            final Account ac = accounts.get(i);
+            current = current.onErrorResume(e -> ac.authenticate());
+        }
+        return current;
     }
 
-    public void login(@Nonnull AuthConfiguration auth) throws LoginFailureException {
-        // update the env state of AzureAccount when auth configuration has a strong configuration of env
-        Objects.requireNonNull(auth, "Null auth configuration is illegal for login.");
-        loginWithAuthConfiguration(auth);
-        if (auth.getEnvironment() != null && this.account.getEnvironment() != null
-                && this.account.getEnvironment() != auth.getEnvironment()
-                && this.account.isAuthenticated()) {
+    public Mono<Account> loginAsync(@Nonnull AuthConfiguration auth) {
+        // update the env state of AzureAccount when auth configuration has an explicit configuration of env
+        Objects.requireNonNull(auth, "Auth configuration is required for login.");
+        Objects.requireNonNull(auth.getType(), "Auth type is required for login.");
+        AuthType type = auth.getType();
+        final List<Account> accounts = new ArrayList<>();
+        if (auth.getType() == AuthType.SERVICE_PRINCIPAL) {
+            accounts.add(new ServicePrincipalAccount(auth));
+        } else {
+            Map<AuthType, Supplier<Account>> accountByType = buildAccountMap();
+            if (auth.getType() == AuthType.AUTO) {
+                accounts.add(new ServicePrincipalAccount(auth));
+                accounts.addAll(accountByType.values().stream().map(Supplier::get).collect(Collectors.toList()));
+            } else {
+                if (!accountByType.containsKey(type)) {
+                    return Mono.error(new LoginFailureException(String.format("Unsupported auth type '%s', supported values are: %s.\"",
+                            type, accountByType.keySet().stream().map(Object::toString).map(StringUtils::lowerCase).collect(Collectors.joining(", ")))));
+                }
+                accounts.add(accountByType.get(type).get());
+            }
+        }
 
-            String expectedEnv = AzureEnvironmentUtils.getCloudNameForAzureCli(auth.getEnvironment());
-            String realEnv = AzureEnvironmentUtils.getCloudNameForAzureCli(this.account.getEnvironment());
+        return authenticateAccounts(accounts).doOnSuccess(result -> {
+            checkEnv(result, auth.getEnvironment());
+            this.setAccount(result);
+        });
+    }
+
+    public Mono<Account> loginAsync(AuthType type) {
```

Comment:
Is `AuthType.AUTO` supported here? Besides, do we need to validate `env` like `loginAsync(@Nonnull AuthConfiguration auth)`

Reply:
auto is not allowed here, I will add a throw exception later